lib/spack/spack/build_environment.py

```diff
@@ -835,154 +831,256 @@ def setup_package(pkg, dirty, context="build"):
+    # Reverse so we go from leaf to root
+    specs_with_type = effective_deptypes(specs, context)
 
-    # Determine 'exe_deps': the set of packages with binaries we want to use
-    if context == "build":
-        exe_deps = build_and_supporting_deps | test_and_supporting_deps
-    elif context == "run":
-        exe_deps = set(spec.traverse(deptype="run"))
-    elif context == "test":
-        exe_deps = test_and_supporting_deps
-
-    def default_modifications_for_dep(dep):
-        if dep in build_link_deps and not is_system_path(dep.prefix) and context == "build":
-            prefix = dep.prefix
-
-            env.prepend_path("CMAKE_PREFIX_PATH", prefix)
-
-            for directory in ("lib", "lib64", "share"):
-                pcdir = os.path.join(prefix, directory, "pkgconfig")
-                if os.path.isdir(pcdir):
-                    env.prepend_path("PKG_CONFIG_PATH", pcdir)
-
-        if dep in exe_deps and not is_system_path(dep.prefix):
-            _make_runnable(dep, env)
-
-    def add_modifications_for_dep(dep):
-        # Some callers of this function only want the custom modifications.
-        # For callers that want both custom and default modifications, we want
-        # to perform the default modifications here (this groups custom
-        # and default modifications together on a per-package basis).
-        if not custom_mods_only:
-            default_modifications_for_dep(dep)
-
-        # Perform custom modifications here (PrependPath actions performed in
-        # the custom method override the default environment modifications
-        # we do to help the build, namely for PATH, CMAKE_PREFIX_PATH, and
-        # PKG_CONFIG_PATH)
-        if dep in custom_mod_deps:
-            dpkg = dep.package
-            if set_package_py_globals:
-                set_module_variables_for_package(dpkg)
-
-            current_module = ModuleChangePropagator(spec.package)
-            dpkg.setup_dependent_package(current_module, spec)
-            current_module.propagate_changes_to_mro()
+    nodes_in_subdag = set(id(s) for s, _ in specs_with_type)
 
-            if context == "build":
-                builder = spack.builder.create(dpkg)
+    # Split into non-external and extenal, maintaining topo order per group.
+    external, nonexternal = stable_partition(reversed(specs_with_type), lambda t: t[0].external)
+
+    should_be_runnable = Mode.BUILDTIME_DIRECT | Mode.RUNTIME_EXECUTABLE
+    should_setup_dependent_build_env = Mode.BUILDTIME | Mode.BUILDTIME_DIRECT
+    should_setup_run_env = Mode.RUNTIME | Mode.RUNTIME_EXECUTABLE
+
+    if context in ("run", "test"):
+        should_setup_run_env |= Mode.ROOT
+        should_be_runnable |= Mode.ROOT
+
+    # I guess everything that calls setup_run_environemnt and setup_dependent_* should have our
+    # horrible globals available too.
+    should_populate_package_py_globals = (
+        should_setup_dependent_build_env | should_setup_run_env | Mode.ROOT
+    )
```

Comment:
(Not necessarily for this PR) I agree that globals are nasty, but people rely on them. An idea to gracefully deprecated them is to:
1. Wrap the globals in some object that can emit a message before doing what the global is expected to do
2. Decide whether how noisy we want to be at the beginning, and how long should we support globals

We can then decide for a "slow" deprecation (e.g. start emitting debug only messages and turn them into warning messages in a couple of releases) or a faster one (warning message, removal in v0.21)